http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1491,182 +1449,359 @@ private <I> Publisher<URI> resolveURI(io.micronaut.http.HttpRequest<I> request,
+                }
+
+                @Override
+                public void finish(ChannelHandlerContext ctx) {
+                    ctx.pipeline().remove(ChannelPipelineCustomizer.HANDLER_MICRONAUT_HTTP_RESPONSE);
+                    if (streamWriter != null) {
+                        if (!streamWriter.isCompleted()) {
+                            // if there was an error, and we didn't fully write the request yet, the
+                            // connection cannot be reused
+                            poolHandle.taint();
+                        }
+                        ctx.pipeline().remove(streamWriter);
+                    }
+                    if (stillExpectingContinue && byteBuf != null) {
+                        byteBuf.release();
+                    }
+                    poolHandle.release();
+                }
+            }));
+            poolHandle.notifyRequestPipelineBuilt();
+
+            HttpHeaders headers = nettyRequest.headers();
+            OptionalLong length = byteBody.expectedLength();
+            if (length.isPresent()) {
+                headers.remove(HttpHeaderNames.TRANSFER_ENCODING);
+                if (length.getAsLong() != 0 || permitsRequestBody(nettyRequest.method())) {
+                    headers.set(HttpHeaderNames.CONTENT_LENGTH, length.getAsLong());
+                }
+            } else {
+                headers.remove(HttpHeaderNames.CONTENT_LENGTH);
+                headers.set(HttpHeaderNames.TRANSFER_ENCODING, HttpHeaderValues.CHUNKED);
+            }
+
+            if (!poolHandle.http2) {
+                if (poolHandle.canReturn()) {
+                    nettyRequest.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.KEEP_ALIVE);
+                } else {
+                    nettyRequest.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+                }
+            }
+
+            Channel channel = poolHandle.channel();
+            if (streamWriter == null) {
+                if (!expectContinue) {
+                    // it's a bit more efficient to use a full request for HTTP/2
+                    channel.writeAndFlush(new DefaultFullHttpRequest(
+                        nettyRequest.protocolVersion(),
+                        nettyRequest.method(),
+                        nettyRequest.uri(),
+                        byteBuf,
+                        nettyRequest.headers(),
+                        EmptyHttpHeaders.INSTANCE
+                    ), channel.voidPromise());
+                } else {
+                    channel.writeAndFlush(nettyRequest, channel.voidPromise());
+                }
+            } else {
+                channel.writeAndFlush(nettyRequest, channel.voidPromise());
+                if (!expectContinue) {
+                    streamWriter.startWriting();
+                }
+            }
+        }).subscribeOn(Schedulers.fromExecutor(poolHandle.channel.eventLoop()));
```

Comment:
does the executor here need to be customisable?

Reply:
No. It is important that the above `Mono.create` lambda runs on the event loop so that the pipeline setup, write calls etc all happen synchronously

Reply:
I will add a comment